File: py-package/tests/test_binary_io.py
import pytest
import numpy as np
import lightgbm as lgb
import lignum as lg
import json

def test_binary_serialization_pipeline(tmp_path):
    np.random.seed(42)
    X = np.random.randint(0, 100, size=(100, 5)).astype(np.float64)
    y = np.random.randint(0, 2, 100).astype(np.float32)
    
    lgb_model = lgb.LGBMClassifier(n_estimators=10, max_depth=3, objective="binary", random_state=42)
    lgb_model.fit(X, y)
    
    json_path = str(tmp_path / "model_transfer.json")
    bin_path = str(tmp_path / "model_transfer.bin")
    
    model_dict = lgb_model.booster_.dump_model()
    with open(json_path, "w") as f:
        json.dump(model_dict, f)
        
    lg_model_json = lg.load(json_path, format="lightgbm")
    
    lg_model_json.save(bin_path)
    lg_model_bin = lg.load(bin_path, format="binary")
    
    expected_raw = lgb_model.predict(X, raw_score=True)
    actual_json_raw = lg_model_json.predict(X, raw_score=True)
    actual_bin_raw = lg_model_bin.predict(X, raw_score=True)
    
    expected_prob = lgb_model.predict_proba(X)[:, 1]
    actual_json_prob = lg_model_json.predict(X, raw_score=False)
    actual_bin_prob = lg_model_bin.predict(X, raw_score=False)
    
    np.testing.assert_allclose(actual_json_raw, expected_raw, rtol=1e-5, atol=1e-5)
    np.testing.assert_allclose(actual_json_prob, expected_prob, rtol=1e-5, atol=1e-5)
    
    np.testing.assert_allclose(actual_bin_raw, actual_json_raw, rtol=1e-7, atol=1e-7)
    np.testing.assert_allclose(actual_bin_prob, actual_json_prob, rtol=1e-7, atol=1e-7)

File: R-package/inst/tinytest/test_binary_io.R
library(tinytest)
library(lignum)

if (requireNamespace("lightgbm", quietly = TRUE)) {
    
  set.seed(42)
  X <- matrix(as.numeric(sample(0:100, 500, replace = TRUE)), nrow = 100, ncol = 5)
  y <- sample(0:1, 100, replace = TRUE)
  
  dtrain <- lightgbm::lgb.Dataset(data = X, label = y)
  params <- list(objective = "binary", max_depth = 3, min_data_in_leaf = 1)
  lgb_model <- lightgbm::lgb.train(params = params, data = dtrain, nrounds = 10, verbose = -1)
  
  json_path <- tempfile(fileext = ".json")
  bin_path <- tempfile(fileext = ".bin")
  
  writeLines(lgb_model$dump_model(), json_path)
  
  lg_model_json <- load_model(json_path, format = "lightgbm")
  
  save_model(lg_model_json, bin_path)
  lg_model_bin <- load_model(bin_path, format = "binary")
  
  expected_raw <- predict(lgb_model, X, type = "raw")
  actual_json_raw <- predict(lg_model_json, X, raw_score = TRUE, n_jobs = 1L)
  actual_bin_raw <- predict(lg_model_bin, X, raw_score = TRUE, n_jobs = 1L)
  
  expected_prob <- predict(lgb_model, X, type = "response")
  actual_json_prob <- predict(lg_model_json, X, raw_score = FALSE, n_jobs = 1L)
  actual_bin_prob <- predict(lg_model_bin, X, raw_score = FALSE, n_jobs = 1L)

  expect_equal(actual_json_raw, expected_raw, tolerance = 1e-5)
  expect_equal(actual_json_prob, expected_prob, tolerance = 1e-5)
  
  expect_equal(actual_bin_raw, actual_json_raw, tolerance = 1e-7)
  expect_equal(actual_bin_prob, actual_json_prob, tolerance = 1e-7)
  
  unlink(json_path)
  unlink(bin_path)
}